src/crimson/os/seastore/seastore_types.h

```diff
@@ -1049,45 +1081,142 @@ class laddr_t {
 };
 
 constexpr laddr_t::Int L_RAW_INT_MAX = std::numeric_limits<laddr_t::Int>::max();
-constexpr laddr_t L_ADDR_MAX = laddr_t(L_RAW_INT_MAX);
+constexpr laddr_t L_ADDR_MAX = laddr_t::from_raw_int(L_RAW_INT_MAX);
 constexpr laddr_t L_ADDR_MIN = laddr_t();
 constexpr laddr_t L_ADDR_NULL = L_ADDR_MAX;
-constexpr laddr_t L_ADDR_ROOT = laddr_t(L_RAW_INT_MAX - 1);
-constexpr laddr_t L_ADDR_LBAT = laddr_t(L_RAW_INT_MAX - 2);
+constexpr laddr_t L_ADDR_ROOT = laddr_t::from_raw_int(L_RAW_INT_MAX - 1);
+constexpr laddr_t L_ADDR_LBAT = laddr_t::from_raw_int(L_RAW_INT_MAX - 2);
+
+struct loffset_t {
+  loffset_t(laddr_t base, extent_len_t offset)
+      : base(base), offset(offset) {}
+
+  laddr_t get_roundup_laddr() const {
+    if (offset == 0) {
+      return base;
+    } else {
+      assert(offset < laddr_t::UNIT_SIZE);
+      return laddr_t(base.value + 1);
+    }
+  }
+  laddr_t get_base() const { return base; }
+  extent_len_t get_offset() const { return offset; }
+
+  operator laddr_t() const {
+    assert(offset == 0);
+    return base;
+  }
+
+  /// static helper functions could access both fields of laddr_t and loffset_t
+  static loffset_t plus(const laddr_t &laddr, extent_len_t offset) {
+    auto base = laddr;
+    base.value += offset >> laddr_t::UNIT_SHIFT;
+    offset &= laddr_t::UNIT_MASK;
+    return loffset_t(base, offset);
+  }
+  static loffset_t minus(const laddr_t &laddr, extent_len_t offset) {
+    auto base = laddr;
+    auto diff = (offset + laddr_t::UNIT_SIZE - 1) >> laddr_t::UNIT_SHIFT;
+    base.value -= diff;
+    offset = (diff << laddr_t::UNIT_SHIFT) - offset;
+    return loffset_t(base, offset);
+  }
+
+  auto operator<=>(const loffset_t&) const = default;
+private:
+  laddr_t base;
+  extent_len_t offset;
+};
+std::ostream &operator<<(std::ostream &out, const loffset_t &loff);
+
+/// =============== operator overload for laddr_t/loffset begin ===============
+inline bool operator==(const laddr_t &laddr, const loffset_t &loff) {
+  return laddr == loff.get_base() && 0 == loff.get_offset();
+}
+inline bool operator==(const loffset_t &loff, const laddr_t &laddr) {
+  return loff.get_base() == laddr && loff.get_offset() == 0;
+}
+
+inline auto operator<=>(const laddr_t &laddr, const loffset_t &loff) {
+  return loffset_t(laddr, 0) <=> loff;
+}
+inline auto operator<=>(const loffset_t &loff, const laddr_t &laddr) {
+  return loff <=> loffset_t(laddr, 0);
+}
+
+inline loffset_t operator+(const laddr_t &laddr, const extent_len_t &off) {
+  return loffset_t::plus(laddr, off);
+}
+inline loffset_t operator+(const loffset_t &loffset, const extent_len_t &length) {
+  return loffset_t::plus(loffset.get_base(), loffset.get_offset() + length);
+}
+inline loffset_t operator+(const extent_len_t &offset, const laddr_t &laddr) {
+  return laddr + offset;
+}
+inline loffset_t operator+(const extent_len_t &length, const loffset_t &loffset) {
+  return loffset + length;
+}
+
+inline laddr_t &operator+=(laddr_t &laddr, extent_len_t len) {
```

Comment:
> After the ObjectDataHandler finishing the construction of extents, the offset of loffset_t is always zero, so I made loffset_t could convert to laddr_t implicitly.

It might be better to drop the implicit conversion, and probably better to write the alignment assertion at the caller side for documentary purpose.